### PR TITLE
layers: Validate geometry shader input topology

### LINKED ISSUE #5651: Validate VUID-VkGraphicsPipelineCreateInfo-pStages-00738/00739
These two VUIDs are currently not validated.

### PR DESCRIPTION
**Pull request by ziga-lunarg**
Closes #5651

### CODE REVIEW

File: layers/core_checks/cc_shader_interface.cpp [hunk 1]
```diff
@@ -207,6 +207,63 @@ bool CoreChecks::ValidateBuiltinLimits(const SPIRV_MODULE_STATE &module_state, c
...
+                                           const StageCreateInfo &create_info, const Location &loc) const {
+    bool skip = false;
+
+    if (!create_info.pipeline || !create_info.pipeline->pre_raster_state) {
+        return skip;
+    }
+
+    const auto &pipeline = *create_info.pipeline;
+
+    if (entrypoint.stage == VK_SHADER_STAGE_GEOMETRY_BIT) {
```
Comment (spencer-lunarg) [inline layers/core_checks/cc_shader_interface.cpp:220]: nit; just early return if not `VK_SHADER_STAGE_GEOMETRY_BIT` like we do with the other conditions above

File: layers/core_checks/cc_shader_interface.cpp [hunk 2]
```diff
@@ -207,6 +207,63 @@ bool CoreChecks::ValidateBuiltinLimits(const SPIRV_MODULE_STATE &module_state, c
...
+        bool has_tess = false;
+        VkPrimitiveTopology topology = pipeline.InputAssemblyState()->topology;
+        for (uint32_t i = 0; i < pipeline.stage_states.size(); i++) {
+            auto &stage_state = pipeline.stage_states[i];
+            const VkShaderStageFlagBits stage = stage_state.GetStage();
+            if (stage == VK_SHADER_STAGE_TESSELLATION_CONTROL_BIT || stage == VK_SHADER_STAGE_TESSELLATION_EVALUATION_BIT) {
```
Comment (spencer-lunarg) [inline layers/core_checks/cc_shader_interface.cpp:226]: should be able to use `pipeline.active_shaders`

File: layers/core_checks/cc_shader_interface.cpp [hunk 3]
```diff
@@ -207,6 +207,63 @@ bool CoreChecks::ValidateBuiltinLimits(const SPIRV_MODULE_STATE &module_state, c
...
+                                 VK_PRIMITIVE_TOPOLOGY_TRIANGLE_FAN, VK_PRIMITIVE_TOPOLOGY_TRIANGLE_LIST_WITH_ADJACENCY,
+                                 VK_PRIMITIVE_TOPOLOGY_TRIANGLE_STRIP_WITH_ADJACENCY}) &&
+            !IsValueIn(geom_topology, {VK_PRIMITIVE_TOPOLOGY_TRIANGLE_LIST, VK_PRIMITIVE_TOPOLOGY_TRIANGLE_STRIP,
+                                       VK_PRIMITIVE_TOPOLOGY_TRIANGLE_FAN, VK_PRIMITIVE_TOPOLOGY_TRIANGLE_LIST_WITH_ADJACENCY,
+                                       VK_PRIMITIVE_TOPOLOGY_TRIANGLE_STRIP_WITH_ADJACENCY});
+        if (mismatch) {
+            if (has_tess) {
+                skip |=
+                    LogError("VUID-VkGraphicsPipelineCreateInfo-pStages-00739", module_state.handle(), loc,
+                             "Geometry shader expects input topology %s, but tessellation evaluation shader output topology is %s.",
```
Comment (spencer-lunarg) [inline layers/core_checks/cc_shader_interface.cpp:254]: these error message I have been starting with

```
"SPIR-V (Geometry stage) expects input ....
```

to help make it obvious it involves the SPIR-V

**Review by spencer-lunarg — CHANGES_REQUESTED** [review]
LGTM, with small fixes

File: layers/core_checks/cc_shader_interface.cpp (hunk 2 above)
Comment (ziga-lunarg) [reply layers/core_checks/cc_shader_interface.cpp:226]: But `pipeline.active_shaders` is just a flag, I need to find the stage state to get the `entrypoint`

**Review by spencer-lunarg — APPROVED** [review]

**Pull request merged**